src/session/components/message_entry.rs

```diff
@@ -0,0 +1,169 @@
+        #[template_child]
+        pub(super) child: TemplateChild<gtk::Overlay>,
+        #[template_child]
+        pub(super) placeholder: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub(super) text_view: TemplateChild<gtk::TextView>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for MessageEntry {
+        const NAME: &'static str = "MessageEntry";
+        type Type = super::MessageEntry;
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            Self::bind_template(klass);
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for MessageEntry {
+        fn properties() -> &'static [glib::ParamSpec] {
+            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> = Lazy::new(|| {
+                vec![glib::ParamSpecBoxed::new(
+                    "formatted-text",
+                    "Formatted text",
+                    "The formatted text of the entry",
+                    BoxedFormattedText::static_type(),
+                    glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
+                )]
+            });
+            PROPERTIES.as_ref()
+        }
+
+        fn set_property(
+            &self,
+            obj: &Self::Type,
+            _id: usize,
+            value: &glib::Value,
+            pspec: &glib::ParamSpec,
+        ) {
+            match pspec.name() {
+                "formatted-text" => obj.set_formatted_text(value.get().unwrap()),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn property(&self, obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            match pspec.name() {
+                "formatted-text" => obj.formatted_text().to_value(),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+
+            self.text_view
+                .buffer()
+                .connect_changed(clone!(@weak obj => move |_| {
+                    obj.text_buffer_changed();
+                }));
+        }
+
+        fn dispose(&self, _obj: &Self::Type) {
+            self.child.unparent();
+        }
+    }
+
+    impl WidgetImpl for MessageEntry {
+        fn grab_focus(&self, _widget: &Self::Type) -> bool {
+            self.text_view.grab_focus()
+        }
+    }
+}
+
+glib::wrapper! {
+    pub(crate) struct MessageEntry(ObjectSubclass<imp::MessageEntry>)
+        @extends gtk::Widget;
+}
+
+impl MessageEntry {
+    pub(crate) fn new() -> Self {
+        glib::Object::new(&[]).unwrap()
+    }
+
+    fn text_buffer_changed(&self) {
+        let imp = self.imp();
+        let buffer = imp.text_view.buffer();
+        let text = buffer
+            .text(&buffer.start_iter(), &buffer.end_iter(), false)
+            .trim()
```

Comment:
The fact that we trim the text means that if we insert just newline in the textview, the placeholder will not disappear, so this has to be fixed.